### ISSUE TITLE
Add missing functionality to names package

### ISSUE THREAD

**Issue by seaneganx** [apigee/registry#446]
### Background
The [names package](https://github.com/apigee/registry/tree/4b43762e57f64242cad80b8c7d2e3add157fde2d/server/registry/names) supports all necessary functionality for the server to work with the names of resources. Its purpose is to abstract away from the strings containing resource names and the details of their expected format so developers can easily work with resource names.

While developing the registry controller, we encountered a situation where we needed to [work with resources names generically](https://github.com/apigee/registry/blob/4b43762e57f64242cad80b8c7d2e3add157fde2d/cmd/registry/controller/parser.go), but the existing names package wasn't designed to work nicely in generic cases. The controller package needed to [extend the existing name types](https://github.com/apigee/registry/blob/4b43762e57f64242cad80b8c7d2e3add157fde2d/cmd/registry/controller/resource-name.go) to fill the gaps.

### Ideas

1. Support full abstraction of resource names: an abstract type that can represent any resource.
2. Support partial abstraction of resource names: abstract types that can represent different kinds of resources (e.g. collections).
3. Add separate types to represent resource collections.
4. Improve identification of parent type when resource names contain a [wildcard](https://google.aip.dev/159#guidance).

**Comment by timburks** [comment]
Would you be open to replacing the word "missing" in the title with "additional"?

**Comment by seaneganx** [comment]
Agreed "missing" isn't the ideal word for it. I don't want to use "Add additional functionality to names package" as a title either though. It doesn't convey that there are gaps between what the package offers and what the needed use cases are.

**Comment by timburks** [comment]
Makes sense, let's describe the missing use case and the impact of addressing it. "Adding generic name parsing to the names package would simplify some places where we parse names" ??
